fix: FutureBuilder in StatelessComponents and deferred imports in non-client side

**Note**: For the entirety of this issue, assume server or static mode and without @client (unless otherwise stated).

**Description**

<!-- A clear and concise description of what the bug is. -->
`FutureBuilder` does not work in `StatelessComponent`s. Regardless of whatever I try, I get this:
```
[SERVER] [ERROR] Asynchronous error
[SERVER] [ERROR] Unsupported operation: Scheduling a frame is not supported on the server, and should never happen.
```
Hence, I cannot use deferred imports like this either:
```dart
import 'package:jaspr/server.dart';

import 'temp.dart' deferred as temp;

// This file is generated automatically by Jaspr, do not remove or edit.
import 'jaspr_options.dart';

void main() {
  Jaspr.initializeApp(
    options: defaultJasprOptions,
  );

  runApp(
    Document(
      title: 'wapik',
      body: App(),
    ),
  );
}

class App extends StatelessComponent {
  const App({super.key});

  @override
  Component build(BuildContext context) {
    return FutureBuilder(
      future: temp.loadLibrary(),
      builder: (context, snapshot) => snapshot.connectionState != ConnectionState.done
          ? text('waiting')
          : text('done'),
    );
  }
}
```

**Steps To Reproduce**

<!-- Please include full steps to reproduce so that we can reproduce the problem. -->

1. Create a server mode or static mode app with default settings and copy paste the above code in main.dart.
2. Create a temp.dart file and put this there:
```dart
const secret = 'von new man';
```
3. Run `jaspr serve`.
4. See error

**Doctor Output**

<!-- Please run `jaspr doctor` and paste the output below. -->
```
[✓] Jaspr CLI (Version 0.21.6)
  • Dart Version 3.9.2 (stable) (Wed Aug 27 03:49:40 2025 -0700) on "linux_x64" at /home/lav/fvm/versions/stable/bin/cache/dart-sdk/bin/dart
  • Running on linux Linux 6.17.5-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 23 Oct 2025 18:49:03 +0000 - Locale en_US.UTF-8
  • Analytics: Enabled
[✓] Current Project
  • Dependencies on core packages:
    • jaspr: ^0.21.6
    • jaspr_builder: ^0.21.6 (dev)
    • jaspr_web_compilers: ^4.2.3+1 (dev)
    • jaspr_router: ^0.7.3
```

**Expected Behavior**

<!-- A clear and concise description of what you expected to happen. -->
No error.

**Additional Context**

<!-- Add any other context about the problem here. -->
The example code is directly inspired by the example in https://docs.jaspr.site/concepts/seo#code-splitting
This is the example there:
```dart
// Use the "deferred as" syntax to specify a deferred import.
import 'components/hello.dart' deferred as hello;

class MyPage extends StatelessComponent {

  @override
  Component build(BuildContext context) {
    return FutureBuilder(
      // This loads in the library when first built.
      future: hello.loadLibrary(),
      builder: (context, snapshot) {
        if (snapshot.hasData) {
          // Show the component only once loaded.
          return hello.Hello();
        } else {
          // Else show a fallback component.
          return Fallback();
        }
      }
    );
  }
}
```

--- a/docs/concepts/seo.mdx
+++ b/docs/concepts/seo.mdx
@@ -84,34 +84,103 @@ Once you have optimized the overall hydration of your site, you can use code spl
 
 Code splitting can improve the load time of your client-side javascript bundle that is compiled from your Dart source. It allows you to split up the compiled javascript into smaller chunks and load them only when needed.
 
+<Info>
+When using [@client](/api/utils/at_client) components in **server** or **static** mode, Jaspr automatically splits code into a separate chunk for each client component. This should be enough for most use-cases and you don't need to perform any additional code-splitting.
+
+If you are using **client** mode only, or want more fine-grained control over code-splitting, read on.
+</Info>
+
 This functionality is built into Dart using a concept called [deferred imports](https://dart.dev/language/libraries#lazily-loading-a-library). You can use a deferred import for lazily importing any Dart library, whether it is in your own codebase or from a package.
 
 For example lazy-loading a component could look like this:
+
+<Tabs defaultValue="client">
+<TabItem label="client mode" value="client">
 
 ```dart
 // Use the "deferred as" syntax to specify a deferred import.
 import 'components/hello.dart' deferred as hello;
 
-class MyPage extends StatelessComponent {
+class MyPage extends StatefulComponent {
+  const MyPage({super.key});
+
+  State<MyPage> createState() => _MyPageState();
+}
+
+class _MyPageState extends State<MyPage> {
+  bool isLoaded = false;
+
+  @override
+  void initState() {
+    super.initState();
+    // Load the library when the component is initialized.
+    hello.loadLibrary().then((_) {
+      setState(() {
+        isLoaded = true;
+      });
+    });
+  }
 
   @override
   Component build(BuildContext context) {
-    return FutureBuilder(
-      // This loads in the library when first built.
-      future: hello.loadLibrary(),
-      builder: (context, snapshot) {
-        if (snapshot.hasData) {
-          // Show the component only once loaded.
-          return hello.Hello();
-        } else {
-          // Else show a fallback component.
-          return Fallback();
-        }
-      }
-    );
+    if (!isLoaded) {
+      // Show a fallback component while loading.
+      return Fallback();
+    }
+
+    // Show the imported component only once loaded.
+    return hello.Hello();
   }
 }
 ```
+
+</TabItem>
+<TabItem label="server/static mode" value="server">
+
+```dart
+// Use the "deferred as" syntax to specify a deferred import.
+import 'components/hello.dart' deferred as hello;
+
+// Mark this (or any ancestor) component with @client to render it on the client (as well as server).
+@client
+class MyPage extends StatefulComponent {
+  const MyPage({super.key});
+
+  State<MyPage> createState() => _MyPageState();
+}
+
+class _MyPageState extends State<MyPage> {
+  bool isLoaded = false;
+
+  @override
+  void initState() {
+    super.initState();
+
+    // Only load the library on the client.
+    if (kIsWeb) {
+      hello.loadLibrary().then((_) {
+        setState(() {
+          isLoaded = true;
+        });
+      });
+    }
+  }
+
+  @override
+  Component build(BuildContext context) {
+    if (!isLoaded) {
+      // Show a fallback component while loading, or on the server during pre-rendering.
+      return Fallback();
+    }
+
+    // Show the imported component only once loaded.
+    return hello.Hello();
+  }
+}
+```
+
+</TabItem>
+</Tabs>
 
 Loading a library like this mostly makes sense if the library is rather complex and would add a significant amount of code to the javascript bundle. Some potential use-cases can be:
 

--- a/packages/jaspr/lib/src/components/async.dart
+++ b/packages/jaspr/lib/src/components/async.dart
@@ -40,6 +40,11 @@ import '../framework/framework.dart';
 ///
 /// `S` is the type of interaction summary.
 ///
+/// When running on the server, `stream` must be null to avoid scheduling unallowed
+/// rebuilds. Use `kIsWeb` to conditionally provide a stream only when running on the web,
+/// or consider using PreloadStateMixin or AsyncStatelessComponent to perform async work
+/// on the server side.
+///
 /// See also:
 ///
 ///  * [StreamBuilder], which is specialized for the case where only the most
@@ -106,6 +111,14 @@ class _StreamBuilderBaseState<T, S> extends State<StreamBuilderBase<T, S>> {
   @override
   void initState() {
     super.initState();
+
+    assert(
+      context.binding.isClient || component.stream == null,
+      'The stream parameter of StreamBuilder must be null when running on the server to avoid scheduling unallowed rebuilds. '
+      'Use `stream: kIsWeb ? yourStream : null` instead, or consider using PreloadStateMixin or '
+      'AsyncStatelessComponent to perform async work on the server side.',
+    );
+
     _summary = component.initial();
     _subscribe();
   }
@@ -324,9 +337,14 @@ typedef AsyncComponentBuilder<T> = Component Function(BuildContext context, Asyn
 ///
 /// Component rebuilding is scheduled by each interaction, using [State.setState],
 /// but is otherwise decoupled from the timing of the stream. The [builder]
-/// is called at the discretion of the Flutter pipeline, and will thus receive a
+/// is called at the discretion of the Jaspr pipeline, and will thus receive a
 /// timing-dependent sub-sequence of the snapshots that represent the
 /// interaction with the stream.
+///
+/// When running on the server, `stream` must be null to avoid scheduling unallowed
+/// rebuilds. Use `kIsWeb` to conditionally provide a stream only when running on the web,
+/// or consider using [PreloadStateMixin] or [AsyncStatelessComponent] to perform async work
+/// on the server side.
 ///
 /// As an example, when interacting with a stream producing the integers
 /// 0 through 9, the [builder] may be called with any ordered sub-sequence
@@ -364,16 +382,6 @@ typedef AsyncComponentBuilder<T> = Component Function(BuildContext context, Asyn
 /// This should be used to ensure that the first frame has the expected value,
 /// as the builder will always be called before the stream listener has a chance
 /// to be processed.
-///
-/// {@tool dartpad}
-/// This sample shows a [StreamBuilder] that listens to a Stream that emits bids
-/// for an auction. Every time the StreamBuilder receives a bid from the Stream,
-/// it will display the price of the bid below an icon. If the Stream emits an
-/// error, the error is displayed below an error icon. When the Stream finishes
-/// emitting bids, the final price is displayed.
-///
-/// ** See code in examples/api/lib/components/async/stream_builder.0.dart **
-/// {@end-tool}
 ///
 /// See also:
 ///
@@ -450,6 +458,11 @@ class StreamBuilder<T> extends StreamBuilderBase<T, AsyncSnapshot<T>> {
 ///
 /// {@youtube 560 315 https://www.youtube.com/watch?v=ek8ZPdWj4Qo}
 ///
+/// When running on the server, `future` must be null to avoid scheduling unallowed
+/// rebuilds. Use `kIsWeb` to conditionally provide a stream only when running on the web,
+/// or consider using [PreloadStateMixin] or [AsyncStatelessComponent] to perform async work
+/// on the server side.
+///
 /// ## Timing
 ///
 /// Component rebuilding is scheduled by the completion of the future, using
@@ -498,17 +511,6 @@ class StreamBuilder<T> extends StreamBuilderBase<T, AsyncSnapshot<T>> {
 /// A [FutureBuilder] behaves identically to a [StreamBuilder] configured with
 /// `future?.asStream()`, except that snapshots with `ConnectionState.active`
 /// may appear for the latter, depending on how the stream is implemented.
-///
-/// {@tool dartpad}
-/// This sample shows a [FutureBuilder] that displays a loading spinner while it
-/// loads data. It displays a success icon and text if the [Future] completes
-/// with a result, or an error icon and text if the [Future] completes with an
-/// error. Assume the `_calculation` field is set by pressing a button elsewhere
-/// in the UI.
-///
-/// ** See code in examples/api/lib/components/async/future_builder.0.dart **
-/// {@end-tool}
-// TODO(ianh): remove unreachable code above once https://github.com/dart-lang/sdk/issues/35520 is fixed
 class FutureBuilder<T> extends StatefulComponent {
   /// Creates a component that builds itself based on the latest snapshot of
   /// interaction with a [Future].
@@ -580,6 +582,14 @@ class _FutureBuilderState<T> extends State<FutureBuilder<T>> {
   @override
   void initState() {
     super.initState();
+
+    assert(
+      context.binding.isClient || component.future == null,
+      'The future parameter of FutureBuilder must be null when running on the server to avoid scheduling unallowed rebuilds. '
+      'Use `future: kIsWeb ? yourFuture : null` instead, or consider using PreloadStateMixin or '
+      'AsyncStatelessComponent to perform async work on the server side.',
+    );
+
     _snapshot = component.initialData == null
         ? AsyncSnapshot<T>.nothing()
         : AsyncSnapshot<T>.withData(ConnectionState.none, component.initialData as T);

